rmc/src/service.rs

```diff
@@ -0,0 +1,420 @@
+//! Reliable MultiCast - a primitive for Reliable Broadcast protocol.
+use crate::{
+    handler::Handler, scheduler::TaskScheduler, RmcHash, RmcIncomingMessage, RmcOutgoingMessage,
+};
+pub use aleph_bft_crypto::{
+    Indexed, MultiKeychain, Multisigned, NodeCount, PartialMultisignature, PartiallyMultisigned,
+    Signable, Signature, Signed, UncheckedSigned,
+};
+use aleph_bft_types::Terminator;
+use core::fmt::Debug;
+use futures::{
+    channel::mpsc::{UnboundedReceiver, UnboundedSender},
+    FutureExt, StreamExt,
+};
+use log::{debug, error, warn};
+use std::hash::Hash;
+
+/// Reliable Multicast Box
+///
+/// The instance of [`Service<H, MK>`] reliably broadcasts hashes of type `H`,
+/// and when a hash is successfully broadcast, the multisigned hash `Multisigned<H, MK>`
+/// is sent to the network.
+///
+/// A node with an instance of [`Service<H, MK>`] can initiate broadcasting a message `msg: H`
+/// by sending the [`RmcIncomingMessage::StartRmc`] variant through the network to [`Service<H, MK>`].
+/// As a result, the node signs `msg` and starts broadcasting the signed message via the network.
+/// When sufficintly many nodes initiate rmc with the same message `msg` and a node collects enough
+/// signatures to form a complete multisignature under the message, the multisigned message
+/// is sent back through the network by the instance of [`Service<H, MK>`].
+///
+/// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/reliable_broadcast.html
+/// for a high-level description of this protocol and how it is used for fork alerts.
+pub struct Service<H: Signable + Hash, MK: MultiKeychain> {
+    network_rx: UnboundedReceiver<RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>>,
+    network_tx: UnboundedSender<RmcOutgoingMessage<H, MK>>,
+    scheduler: Box<dyn TaskScheduler<RmcHash<H, MK::Signature, MK::PartialMultisignature>>>,
```

Comment:
No need for `dyn` here. You can use a generic parameter (some more juggling with `TestEnvironment`) or just `Double...` since its the only implementation anyway.

Reply:
yeah, that's much cooler